win32/tests/file_win32_ut/file_win32_ut.c

```diff
@@ -0,0 +1,952 @@
+        .SetReturn(FALSE);
+    STRICT_EXPECTED_CALL(mock_GetLastError())
+        .SetReturn(ERROR_IO_INCOMPLETE);
+    STRICT_EXPECTED_CALL(mock_CancelThreadpoolIo(IGNORED_ARG))
+        .ValidateArgumentValue_pio(&captured_ptp_io);
+    STRICT_EXPECTED_CALL(mock_CloseHandle(h_event));
+    STRICT_EXPECTED_CALL(free(IGNORED_ARG))
+        .ValidateArgumentValue_ptr(&io);
+    ///act
+    FILE_WRITE_ASYNC_RESULT result = file_write_async(file_handle, source, size, position, user_callback, NULL);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_WRITE_ERROR, result);
+    
+    ///cleanup
+    file_destroy(file_handle);
+}
+
+/*Tests_SRS_FILE_WIN32_43_017: [ file_write_async shall call StartThreadpoolIo.]*/
+/*Tests_SRS_FILE_WIN32_43_054: [ file_write_async shall create an event by calling CreateEvent.]*/
+/*Tests_SRS_FILE_WIN32_43_020: [ file_write_async shall allocate an OVERLAPPED struct and populate it with the created event and position.]*/
+/*Tests_SRS_FILE_WIN32_43_018: [ file_write_async shall allocate a context to store the allocated OVERLAPPED struct, handle, size, user_callback and user_context. ]*/
+/*Tests_SRS_FILE_WIN32_43_021: [ file_write_async shall call WriteFile with handle, source, size and the allocated OVERLAPPED struct.]*/
+/*Tests_SRS_FILE_WIN32_43_024: [ If WriteFile succeeds synchronously then file_write_async shall succeed, call CancelThreadpoolIo, call user_callback and return FILE_WRITE_ASYNC_OK. ]*/
+TEST_FUNCTION(file_write_async_succeeds_synchronously)
+{
+    ///arrange
+    PTP_WIN32_IO_CALLBACK captured_callback;
+    FILE_HANDLE file_handle = get_file_handle("file_write_async_succeeds_synchronously.txt", &captured_callback);
+    unsigned char source[10];
+    uint32_t size = 10;
+    uint64_t position = 5;
+    HANDLE h_event = (HANDLE)44;
+    void* user_context = (void*)45;
+    void* io;
+    PTP_IO captured_ptp_io;
+
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG))
+        .CaptureReturn(&io);
+    STRICT_EXPECTED_CALL(mock_CreateEvent(IGNORED_ARG, IGNORED_ARG, IGNORED_ARG, IGNORED_ARG))
+        .SetReturn(h_event);
+    STRICT_EXPECTED_CALL(mock_StartThreadpoolIo(IGNORED_ARG))
+        .CaptureArgumentValue_pio(&captured_ptp_io);
+    STRICT_EXPECTED_CALL(mock_WriteFile(IGNORED_ARG, source, size, NULL, IGNORED_ARG))
+        .SetReturn(TRUE);
+    STRICT_EXPECTED_CALL(mock_CancelThreadpoolIo(IGNORED_ARG))
+        .ValidateArgumentValue_pio(&captured_ptp_io);
+    STRICT_EXPECTED_CALL(mock_user_callback(user_context, true));
+    STRICT_EXPECTED_CALL(mock_CloseHandle(h_event));
+    STRICT_EXPECTED_CALL(free(IGNORED_ARG))
+        .ValidateArgumentValue_ptr(&io);
+    
+    ///act
+    FILE_WRITE_ASYNC_RESULT result = file_write_async(file_handle, source, size, position, (FILE_CB)mock_user_callback, user_context);

```

Comment:
>FILE_CB [](start = 92, length = 7)

If you are casting a callback type that tells me something is wrong and will crash horribly at some point when you least expect it...or the cast isn't needed.